Extended documentation for `transformations.jl` (#73)
bsc-quantic/Tenet.jl: docs/src/transformations.md

# Transformations

> In tensor network computations, it is common to apply various transformations to simplify the network structure, reduce computational cost, or prepare the network for further operations. These transformations can modify the network's structure in a variety of ways such as reducing dimensions or ranks of tensors, converting certain indices, or even reordering tensor indices.

Reviewer: Actually is not common, but its a good idea.
```suggestion
In tensor network computations, it is good practice to apply various transformations to simplify the network structure, reduce computational cost, or prepare the network for further operations. These transformations modify the network's structure locally by permuting, contracting, factoring or truncating tensors.
```

A crucial reason why these methods are indispensable lies in their ability to drastically reduce the problem size of the contraction path search and also the contraction. This doesn't necessarily involve reducing the maximum rank of the Tensor Network itself, but more importantly, it reduces the size (or rank) of the involved tensors.

Our approach has been significantly inspired by the ideas presented in the [Quimb](https://quimb.readthedocs.io/) library, explained in [this paper](https://arxiv.org/pdf/2002.01935.pdf).

In Tenet, we provide a set of predefined transformations which you can apply to your `TensorNetwork` using both the `transform`/`transform!` functions.

```@docs
transform
transform!
```

# Example
Here we show how can we reduce the complexity of the tensor network by applying a tranformation to it:

> ```julia-repl
> julia> tn = TensorNetwork(...)
> julia> transformed_tn = transform(tn, Tenet.RankSimplification)
> 
> julia> fig = Figure()
> julia> ax1 = Axis(fig[1, 1]; title="Original TensorNetwork")
> julia> p1 = plot!(ax1, tn; node_size=5.)
> julia> ax2 = Axis(fig[1, 2], title="Transformed TensorNetwork")
> julia> p2 = plot!(ax2, tn2; node_size=5.)
> julia> ax1.titlesize=20
> julia> ax2.titlesize=20
> ```
> ```@raw html
> <figure>
> <img width=500 src="../assets/transformation.svg" alt="Before and after transformation in a Tensor Network">
> </figure>

Reviewer: Replace this with a `@example` block so the plot is generated during the documentation generation process.
```suggestion
```@example
tn = TensorNetwork(...)
transformed_tn = transform(tn, Tenet.RankSimplification)

fig = Figure() # hide
ax1 = Axis(fig[1, 1]; title="Original TensorNetwork") # hide
p1 = plot!(ax1, tn; node_size=5.) # hide
ax2 = Axis(fig[1, 2], title="Transformed TensorNetwork") # hide
p2 = plot!(ax2, tn2; node_size=5.) # hide
ax1.titlesize=20 # hide
ax2.titlesize=20 # hide
fig # hide
```

Author: But then I would need to load the qasm circuit using `QuacIO`

# Transformations

In tensor network computations, it is common to apply various transformations to simplify the network structure, reduce computational cost, or prepare the network for further operations. These transformations can modify the network's structure in a variety of ways such as reducing dimensions or ranks of tensors, converting certain indices, or even reordering tensor indices.

> In Tenet, we provide a set of predefined transformations which you can apply to your `TensorNetwork` using both the `transform` and `transform!` functions. The `transform` function is non-mutating, meaning it will return a new `TensorNetwork` where the transformation has been applied, leaving the original network unchanged. For in-place transformation, you can use the mutating version, `transform!`.

Reviewer: ```suggestion
In Tenet, we provide a set of predefined transformations which you can apply to your `TensorNetwork` using both the `transform`/`transform!` functions. The `transform` function is non-mutating, meaning it will return a new `TensorNetwork` where the transformation has been applied, leaving the original network unchanged. For in-place transformation, you can use the mutating version, `transform!`.
```

> In Tenet, we provide a set of predefined transformations which you can apply to your `TensorNetwork` using both the `transform` and `transform!` functions. The `transform` function is non-mutating, meaning it will return a new `TensorNetwork` where the transformation has been applied, leaving the original network unchanged. For in-place transformation, you can use the mutating version, `transform!`.

Reviewer: This already known by using the `!` suffix.

```suggestion
In Tenet, we provide a set of predefined transformations which you can apply to your `TensorNetwork` using both the `transform` and `transform!` functions.
```

Reviewer: @jofrevalles what about removing these 2 phrases?

```@docs
transform
transform!
```

# Examples
```julia-repl
julia> tn = TensorNetwork(...)
julia> transformed_tn = transform(tn, Tenet.RankSimplification)

julia> fig = Figure()
julia> ax1 = Axis(fig[1, 1]; title="Original TensorNetwork")
julia> p1 = plot!(ax1, tn; node_size=5.)
julia> ax2 = Axis(fig[1, 2], title="Transformed TensorNetwork")
julia> p2 = plot!(ax2, tn2; node_size=5.)
julia> ax1.titlesize=20
julia> ax2.titlesize=20
```
```@raw html
<figure>
<img width=500 src="../assets/transformation.svg" alt="Before and after transformation in a Tensor Network">
</figure>
```

> 

Reviewer: I prefer to put examples inside the docstrings of the different `Transformation`s, so the user can check their effect.

(Use small examples)

Author: Should we show a visual representation for each transformation? Maybe this would be too much

Reviewer: I think this would be a good idea, but using very small tensor networks. Similar to what they do in the paper with the diagrams.

Reviewer: I can help with this if you need help.

Author: I will try to create some examples, I will let you know!

# Transformations

```@docs
Tenet.HyperindConverter

> ```
> 
> ```@docs
> Tenet.DiagonalReduction
> ```
> ```@docs
> Tenet.RankSimplification
> ```
> 
> ```@docs
> Tenet.AntiDiagonalGauging
> ```
> 
> ```@docs
> Tenet.ColumnReduction
> ```
> 
> ```@docs
> Tenet.SplitSimplification
> ```

Reviewer: Why this separation?
```suggestion
Tenet.DiagonalReduction
Tenet.RankSimplification
Tenet.AntiDiagonalGauging
Tenet.ColumnReduction
Tenet.SplitSimplification
```
